### PR TITLE
Rate limit streams [DPP-1001]

### PR DESCRIPTION
**Pull request by simonmaxen-da**

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 1]
```diff
@@ -65,30 +74,39 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
       (for {
-        _ <- metricOverloaded(fullMethodName, apiServices, config.maxApiServicesQueueSize)
+        _ <- queueOverloaded(apiServices, config.maxApiServicesQueueSize)
+        _ <- queueOverloaded(indexDbThreadpool, config.maxApiServicesIndexDbQueueSize)
         _ <- metricOverloaded(
-          fullMethodName,
-          indexDbThreadpool,
-          config.maxApiServicesIndexDbQueueSize,
+          metricDescription = "Number of active streams",
+          name = metrics.daml.lapi.streams.activeName,
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:81]: could we get the reference to `metrics.daml.lapi.streams` only once?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 2]
```diff
@@ -39,23 +41,30 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
+
         new ServerCall.Listener[ReqT]() {}
 
       case None =>
-        next.startCall(call, headers)
+        val listener: ServerCall.Listener[ReqT] = next.startCall(
+          new OnCloseServerCall(call, () => metrics.daml.lapi.streams.active.dec()),
+          headers,
+        )
+        metrics.daml.lapi.streams.active.inc() // Only do after call above has returned
```
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:66]: You could go <0 in extreme situation, but otherwise it would be harmless

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 3]
```diff
@@ -98,9 +116,27 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
-  val doNonLimit = Set(
+  val doNonLimit: Set[String] = Set(
     "grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo",
     "grpc.health.v1.Health/Check",
     "grpc.health.v1.Health/Watch",
   )
+
+  private class OnCloseServerCall[ReqT, RespT](
+      val delegate: ServerCall[ReqT, RespT],
+      safeOnClose: () => Unit,
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:207]: Because exceptions will be cause and logged (not thrown).

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 4]
```diff
@@ -39,23 +41,30 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
         val exception = StatusProto.toStatusRuntimeException(rpcStatus)
 
         call.close(exception.getStatus, exception.getTrailers)
+
         new ServerCall.Listener[ReqT]() {}
 
       case None =>
-        next.startCall(call, headers)
+        val listener: ServerCall.Listener[ReqT] = next.startCall(
+          new OnCloseServerCall(call, () => metrics.daml.lapi.streams.active.dec()),
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:73]: Not possible as this is calling into java code.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:76]: Agree.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 5]
```diff
@@ -65,30 +74,39 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
-          """.stripMargin
+    metricOverloaded(
+      metricDescription = s"${count.name} queue size",
+      name = count.prefix,
+      value = count.queueSize,
+      limit = limit,
+    )
+  }
 
+  private def metricOverloaded(
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:157]: I think the return types should stay the same as it makes the composition work.  It may make more sense to assert that things are under limit as opposed to over limit though.  (e.g. `memoryUnderLimit`).

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/RateLimitingConfig.scala [hunk 6]
```diff
@@ -6,11 +6,13 @@ package com.daml.platform.apiserver.configuration
 final case class RateLimitingConfig(
     maxApiServicesQueueSize: Int,
     maxApiServicesIndexDbQueueSize: Int,
+    maxStreams: Int,
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/RateLimitingConfig.scala:25]: The second two have been renamed.

When it comes to the first:
Everything is 'Api' => maxActiveStreamingCalls
This will limit both active _and_ inactive calls => maxStreamingCalls
Everything is a call => maxStreams   ;-)

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 7]
```diff
@@ -65,30 +74,39 @@ private[apiserver] final class RateLimitingInterceptor(metrics: Metrics, config:
...
-        _ <- metricOverloaded(fullMethodName, apiServices, config.maxApiServicesQueueSize)
+        _ <- queueOverloaded(apiServices, config.maxApiServicesQueueSize)
+        _ <- queueOverloaded(indexDbThreadpool, config.maxApiServicesIndexDbQueueSize)
         _ <- metricOverloaded(
-          fullMethodName,
-          indexDbThreadpool,
-          config.maxApiServicesIndexDbQueueSize,
+          metricDescription = "Number of active streams",
+          name = metrics.daml.lapi.streams.activeName,
+          value = metrics.daml.lapi.streams.active.getCount + 1, // Once created
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:93]: `active.getCount + 1` will be the number of active streams if we don't rate limit this call - update comment.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala [hunk 8]
```diff
@@ -164,6 +168,82 @@ final class RateLimitingInterceptorSpec
     }
   }
 
+  it should "limit the number of streams" in {
+
+    val metrics = new Metrics(new MetricRegistry)
+
+    val limitStreamConfig = RateLimitingConfig.default.copy(maxStreams = 2)
+
+    class Stream(val close: () => Future[Status], val status: Future[Status])
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:232]: As it seems calling `request` and `close` on a closed ServiceCall does throw an error it is no longer needed (removed)

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala [hunk 9]
```diff
@@ -164,6 +168,82 @@ final class RateLimitingInterceptorSpec
...
+            if (Set(1, size).contains(message.respInt)) logger.info(s"onMessage($message)")
+            if (message.respInt == 1) init.success(stream)
+          }
+        },
+        new Metadata(),
+      )
+
+      // When the handshake is single message -> stream then onReady is not applicable
+      clientCall.sendMessage(HelloRequest(size, ByteString.copyFromUtf8(id)))
+      clientCall.halfClose()
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:266]: Doing exactly this, see `WaitService`.

**Comment by simonmaxen-da** [comment]
LR passes, see:
 https://lr.da-ext.net/grafana/d/sandbox-on-x/sandbox-on-x?orgId=1&var-benchmark_id=8631&from=1655386333000&to=1655387585000

File: ledger-api/sample-service/src/main/scala/com/digitalasset/grpc/sampleservice/implementations/HelloServiceReferenceImplementation.scala [hunk 10]
```diff
@@ -26,6 +26,7 @@ class HelloServiceReferenceImplementation
       request: HelloRequest,
       responseObserver: StreamObserver[HelloResponse],
   ): Unit = {
+
```
Comment (mziolekda) [inline ledger-api/sample-service/src/main/scala/com/digitalasset/grpc/sampleservice/implementations/HelloServiceReferenceImplementation.scala:29]: Intentional?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 11]
```diff
@@ -186,11 +213,39 @@ object RateLimitingInterceptor {
...
 
+  private val UnderLimit: Either[String, Unit] = Right(())
+
   val doNonLimit: Set[String] = Set(
     "grpc.reflection.v1alpha.ServerReflection/ServerReflectionInfo",
     "grpc.health.v1.Health/Check",
     "grpc.health.v1.Health/Watch",
   )
+
+  private class OnCloseCallListener[RespT](
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:224]: This file is slowly becoming hard to follow. I would consider spliting different unrelated classes into their own files in this module

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 12]
```diff
@@ -186,11 +213,39 @@ object RateLimitingInterceptor {
     )
   }
 
+  private val UnderLimit: Either[String, Unit] = Right(())
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:216]: With this definition, we are one step away from an ADT, why not go all the way?

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala [hunk 13]
```diff
@@ -217,6 +223,126 @@ final class RateLimitingInterceptorSpec
...
+
+    val waitService = new WaitService()
+    withChannel(metrics, waitService, limitStreamConfig).use { channel: Channel =>
+      {
+        for {
+
+          fStatus1 <- streamHello(channel)
+          fHelloStatus1 = singleHello(channel)
+          fStatus2 <- streamHello(channel)
+          fHelloStatus2 = singleHello(channel)
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:275]: I would add another `streamHello` right above to ascertain that a stream is rejected where a single is allowed

Comment (mziolekda) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:275]: Shouldn't we get the metric at this point? How else do we know that the counter has not gone up?

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala [hunk 14]
```diff
@@ -217,6 +223,126 @@ final class RateLimitingInterceptorSpec
...
+    val limitStreamConfig = RateLimitingConfig.Default.copy(maxStreams = 2)
+
+    val waitService = new WaitService()
+    withChannel(metrics, waitService, limitStreamConfig).use { channel: Channel =>
+      {
+        for {
+
+          fStatus1 <- streamHello(channel)
+          fStatus2 <- streamHello(channel)
+          fHelloStatus1 = singleHello(channel)
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:311]: I would add another `streamHello` right above to show that streams are limited while singles aren't

File: ledger-api/sample-service/src/main/scala/com/digitalasset/grpc/sampleservice/implementations/HelloServiceReferenceImplementation.scala (hunk 10 above)
Comment (simonmaxen-da) [reply ledger-api/sample-service/src/main/scala/com/digitalasset/grpc/sampleservice/implementations/HelloServiceReferenceImplementation.scala:29]: No.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala (hunk 13 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:275]: Yes, makes it more explicit, especially given the test name.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala (hunk 14 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:311]: The "limit the number of streams" test above does exactly that already.  Repeating that test here may detract from the purpose of this test.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala (hunk 11 above)
Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:241]: I've created new branch `rate-limit-refactor` that makes things far more manageable. See https://github.com/digital-asset/daml/pull/14311 PR for description.

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala [hunk 15]
```diff
@@ -54,7 +62,32 @@ final class RateLimitingInterceptorSpec
...
+  private val config = RateLimitingConfig(100, 10, 75, 100 * RateLimitingConfig.Megabyte, 100)
+
+  behavior of "LimitResult"
+
+  def underCheck(): LimitResult = UnderLimit
+  def overCheck(method: String): LimitResult = OverLimit(
+    MaximumNumberOfStreams.Rejection(0, 0, "", method)(NoLogging)
+  )
+
+  it should "compose under limit" in {
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:74]: actually, ignore my comment, `LimitResult` is not generic over some time, so just not applicable here.

**Review by skisel-da — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala [hunk 16]
```diff
@@ -133,26 +162,53 @@ private[apiserver] final class RateLimitingInterceptor(
...
+          limit = limit,
+          metricPrefix = count.prefix,
+          fullMethodName = fullMethodName,
+        )
+      )
+    } else UnderLimit
+  }
+
+  private def streamsUnderLimit(fullMethodName: String): LimitResult = {
+    if (activeStreamsCounter.getCount >= config.maxStreams) {
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:181]: if activeStreamsCounter is == maxStreams, it should still be fine no?

Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:181]: But then this also really limits the max number of subscriptions to maxStreams - 1, potentially confusing users, no?

Comment (simonmaxen-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/RateLimitingInterceptor.scala:181]: The rate limiting test is applied before the count (for the stream being created) is incremented.

At the point the `maxStream`'th stream is being created `activeStreamsCounter` will contain `maxStream-1` (and so not be limited).  (the counter is incremented post rate limiting check).

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala (hunk 14 above)
Comment (mziolekda) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/RateLimitingInterceptorSpec.scala:342]: :+1:

**Review by mziolekda — APPROVED** [review]
